hybrid/compartment_examples/inter_comp_call_secure/main.c

```diff
@@ -0,0 +1,194 @@
+
+/* Abstract representation of a compartment. Within 80 bytes we represent:
+ * - an id (8B)
+ * - address to the start of the compartment's stack (8B)
+ *   the size of the stack (8B)
+ * - address to the start of the compartment's heap (8B)
+ *   the size of the heap (8B)
+ * - alignment padding (8B)
+ * - the ddc corresponding to the compartment (16B)
+ * - the function within the compartment to be executed upon switching (16B)
+ */
+struct comp
+{
+	size_t id;
+	void *stack_addr;
+	size_t stack_len;
+	void *heap_addr;
+	size_t heap_len;
+	void *__capability ddc;
+	void *__capability comp_fn;
+};
+
+// ASM offsets, included here for validation
+#include "main.h"
+
+static_assert(COMP_SIZE == sizeof(struct comp), "Invalid `COMP_SIZE` provided");
+static_assert(COMP_OFFSET_STK_ADDR == offsetof(struct comp, stack_addr),
+			  "Invalid `COMP_OFFSET_STK_ADDR` provided.");
+static_assert(COMP_OFFSET_STK_LEN == offsetof(struct comp, stack_len),
+			  "Invalid `COMP_OFFSET_STK_LEN` provided.");
+static_assert(COMP_OFFSET_DDC == offsetof(struct comp, ddc), "Invalid `COMP_OFFSET_DDC` provided.");
+static_assert(COMP_OFFSET_PCC == offsetof(struct comp, comp_fn),
+			  "Invalid `COMP_OFFSET_PCC` provided.");
+
+struct comp comps[COMP_COUNT];
+
+/*******************************************************************************
+ * Privileged Functions
+ ******************************************************************************/
+
+void init_comps()
+{
+	void *__capability switch_cap = (void *__capability) switch_compartment;
+	switch_cap = cheri_bounds_set(switch_cap, 80 * 4);
+	asm("seal %w0, %w0, rb" : "+r"(switch_cap) :);
+	saved_caps[0] = switch_cap;
+
+	void *__capability comps_addr = (void *__capability) &comps;
+	comps_addr = cheri_bounds_set(comps_addr, COMP_COUNT * COMP_SIZE);
+	saved_caps[1] = comps_addr;
+}
+
+/* This function stores the compartment data in memory, so that it can be
+ * accessed by the compartments as needed.
+ */
+void executive_switch(struct comp c)
+{
+	void *__capability comps_addr = saved_caps[1];
+
+	asm("mov c19, %w0\n\t"
+		"mov x0, #0\n\t"
+		"msr CID_EL0, c0"
+		:
+		: "r"(comps_addr));
+
+	void *__capability chk = *((void *__capability *) (comps[0].heap_addr));
+	assert(cheri_is_valid(chk));
+	switch_compartment();
+}
+
+/* Store capabilities we want saved within each compartment */
+void update_comp_saved_caps(void *__capability *addr)
+{
+	for (size_t i = 0; i < saved_caps_count; ++i)
+	{
+		addr[i] = saved_caps[i];
+	}
+}
+
+void add_comp(uint8_t *_start_addr, void (*_comp_fn)())
+{
+	assert(id < COMP_COUNT);
+
+	struct comp new_comp;
+	new_comp.id = id;
+
+	new_comp.stack_addr = (void *) (_start_addr + total_comp_size - comp_stack_size);
+	new_comp.stack_len = comp_stack_size;
+	new_comp.heap_addr = (void *) _start_addr;
+	new_comp.heap_len = total_comp_size - comp_stack_size;
+
+	// Memory magic; we will store capabilities at the memory location pointed
+	// by `heap_addr` of current compartment
```

Comment:
Wouldn't that corrupt the heap?

Reply:
I'm unsure what corrupt the heap implies, but it is the case that we consider the heap region where we store this info as "reserved", and ideally a memory allocator would avoid overwriting this.

Reply:
I think that's my point: if the allocator avoids overwriting it, then it's not really part of the heap, and instead is a region in its own right.
Note that this won't support re-entrant or multi-threaded compartments that might need to have multiple calls active at any one time, but that's Ok here as long as the limitation is noted.

Reply:
That seems to be a problem to consider when we implement memory management for these compartments, in however fashion that might be. For this example, I believe what we call "compartment heap" has no actual meaning, so, if we do mention it (as I believe I have in the intro comment), it should be fine?

Reply:
I'm arguing only for calling it something else, since this register storage area is fundamental to the switching mechanism. The actual behaviour is fine, as you point out, but this isn't a "heap_addr" (yet).

Reply:
Ah, I think I missed to explain a point. It just so happens that the memory location I want to place this capability is contained within the `heap_addr` variable, which determines where the compartment's highest heap address is. That's why I'm using this variable, named `heap_addr` in the `struct`.

Actually I just realised I made a mistake in the memory design of the compartments. In the diagram [1], I have lower addresses at the top, when I meant to have higher ones. The idea is to store this capability at the highest heap address, far away from where user-affected memory would lie.

[1] https://github.com/capablevms/cheri-examples/pull/55/files#diff-1ba17d80ea42080e89335c8f51d44a66e80fe3df28b74490c2f8e4a36ac1c786R28